### PR TITLE
Sidenote using footnote syntax

### PR DESCRIPTION
**Pull request by AakashGfude**
Taking @mmcky and @choldgraf's comments into account. Implemented the sidenote and marginnote using footnote syntax as the first step. Changing footnote nodes to sidenote and inline element nodes in a post-transform. This does not create any new syntax to write sidenotes and uses the already available footnote syntax. To convert all footnotes to sidenotes, the config variable is - `use_sidenotes: True` under `html_theme_options`. To convert a sidenote to marginnote (i.e. not haveing superscripted numbers), use `{>}` at the start of the footnote source text (similar syntax was first proposed in https://github.com/executablebooks/sphinx-book-theme/pull/523#issuecomment-1069154993). 

**Example**:

 _Source Code:_
```
On wide screens, margin content will pop out to the side of the page and allow content underneath it to move upwards[^myref].
This allows you to provide extra information without interrupting the flow of information.

There are many other things that you can do with the Sphinx Book Theme.
Now that you've gotten a start, check out the other sections to the left to learn more about how to use it[^3].

[^myref]: Some footnote text

[^3]: And here's a longer margin block-level element...
      :::{note}
      ...with a note inside.
      :::
```

_Footnote:_

Screen Shot 2022-04-06 at 4 55 48 pm

_Sidenote:_

Screen Shot 2022-04-06 at 4 55 18 pm

When modifying the above source code with:

```
[^myref]:  {>} Some footnote text
```

The output is:

Screen Shot 2022-04-07 at 10 53 04 am

Note that sidenotes do not support any directives inside it as it is an inline element. Tufte also uses inline span elements. The reason span makes more sense than aside in my opinion is mentioned in the **31st March** description below.

**NOTE**: The functionality of clickable text/icon on smaller screens as was implemented in https://github.com/executablebooks/sphinx-book-theme/pull/523, is also present in this PR.

## 31st march

Was trying to implement sidenote leveraging the footnote code and syntax as discussed in https://github.com/executablebooks/sphinx-book-theme/pull/523 . I tried going through the docutils code, where the footnote is implemented. 
I have a few concerns which I wanted to discuss:

- the insertion of `footnote` node in the document happens during the parsing phase I think. And possibly the only place to rearrange the document to make the `footnote` node move near the `footnote-reference` is during a post-transform. 

- It was much easier to deal with `span` elements we generated here https://github.com/executablebooks/sphinx-book-theme/pull/523 for sidenotes, and which Tufte CSS uses as well, to position the sidenote. As `span` can go inside a `p` tag, and other similar tags, which will have the sidenote reference inside it. And so relative positioning is pretty straightforward. The `aside` tag on the other hand, when trying to insert inside a 'p' tag in post-transform, does not work, and it is always inserted after the `p` tag, which makes it hard to position. 

cc: @choldgraf @chrisjsewell @mmcky @rowanc1

### CODE REVIEW

**Comment by mmcky** [comment]
@AakashGfude I have been taking a look at this PR in addition to #523 and from a `developer` perspective it seems to me that it would make sense to:

1. Develop the concept of a `sidenote` as a feature with supporting `visit` and `depart` methods for `html` and `latex`
2. Add a global option `footnote_as_sidenote: [True/False]` which enables a `sphinx.transform` to move `footnote` objects (visually) to `sidenote` objects in the `AST` to be parsed by `sidenote` writers. 

To me this flow seems more logical rater than overloading the `footnote` objects to create `sidenote` style of output. Some of the benefits of this approach are improvements made to `sidenote` feature are automatically gained if we use a `transform` to move `footnotes` to `sidenote` nodes in the AST.

**Comment by AakashGfude** [comment]
Thanks, @mmcky, that sounds like a reasonable way to go. I will approach it that way and see how it feels.

**Comment by AakashGfude** [comment]
made changes, and updated the top description.

**Review by choldgraf — COMMENTED** [review]
This is looking nice! Much cleaner than having lots of new roles/directives, I think. A few quick thoughts:

- I pushed a commit that added the extra docs from your other PR, and updated them a bit for this new syntax.
- I updated the "marginnote" prefix to `{-}` instead of `{>}` to be in-line with what the pandoc-sidenotes extension uses.
- I've added some example usage in our reference docs as well so we can show off the behavior there.
- I noticed that having multiple marginnotes in the same sentence does not work on mobile screens, I think it's because the `for` is pointing to the same thing in both cases, can you fix that?
- I think that it'd be good to have some visual cue that the numbers/symbols are "clickable" on mobile - maybe they can still be blue? Or some other way to suggest they can be clicked?

**Comment by AakashGfude** [comment]
Thank you so much @choldgraf, for the additions, very useful. 

>     * I noticed that having multiple margin notes in the same sentence does not work on mobile screens, I think it's because the `for` is pointing to the same thing in both cases, can you fix that?

Have pushed the fix now.

>     * I think that it'd be good to have some visual cue that the numbers/symbols are "clickable" on mobile - maybe they can still be blue? Or some other way to suggest they can be clicked?

Yes, I think the original color (blue) seems like a good option. As the user is already used to assuming that it is a clickable reference? Have pushed the change. Let's see how it feels.

**Comment by choldgraf** [comment]
I think the current implementation is pretty good to me. The behavior is nearly identical to how [pandoc-sidenotes](https://github.com/jez/pandoc-sidenote) works, so we're not reinventing the wheel from a UX perspective or anything.

One thing I noticed: If you have a marginnote in between two sidenotes, it still takes up a "number" even though it is not actually numbered (so the sidenotes will jump from 1 to 3 even though there is no 2). Is that easy to fix? If not it's not a blocker on this PR in my opinion.

File: src/sphinx_book_theme/nodes.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        add_node(
+            cls,
+            override=True,
+            html=(visit_SideNoteNode, depart_SideNoteNode),
+            latex=(visit_SideNoteNode, depart_SideNoteNode),
+        )
+
+
+def visit_SideNoteNode(self, node):
+    if isinstance(self, LaTeXTranslator):
```
Comment (chrisjsewell) [inline src/sphinx_book_theme/nodes.py:25]: The post transform should be set to only run on the html format, then you can get rid of this

**Review by chrisjsewell — CHANGES_REQUESTED** [review]

File: src/sphinx_book_theme/nodes.py (hunk 1 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/nodes.py:25]: sounds good. The way to handle it in latex should probably be in `sphinx-jupyterbook-latex`? At present, the pdf just ignores margin texts. 
Unless we decide to move this functionality out of `sphinx-book-theme`

Comment (chrisjsewell) [reply src/sphinx_book_theme/nodes.py:25]: Yep exactly 👍, whilst it is only in sphinx-book-theme, it should only affect html

**Comment by AakashGfude** [comment]
> One thing I noticed: If you have a marginnote in between two sidenotes, it still takes up a "number" even though it is not actually numbered (so the sidenotes will jump from 1 to 3 even though there is no 2). Is that easy to fix? If not it's not a blocker on this PR in my opinion.

ooh yeah, it's because we are converting footnotes to sidenotes/margin notes and using the numbering of footnotes. margin notes were also footnotes and had a number.

File: src/sphinx_book_theme/nodes.py (hunk 1 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/nodes.py]: cool, have added check for `standaloneHTMLBuilder`.

File: src/sphinx_book_theme/_transforms.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+from .nodes import SideNoteNode
+from sphinx.builders.html import StandaloneHTMLBuilder
+
+
+class HandleFootnoteTransform(SphinxPostTransform):
+
+    default_priority = 1
+
+    def apply(self, **kwargs: Any) -> None:
+        if isinstance(self.env.app.builder, StandaloneHTMLBuilder):
```
Comment (chrisjsewell) [inline src/sphinx_book_theme/_transforms.py:13]: No you just add `formats = ("html",)` (have a look at the sphinx post transform documentation)

**Review by chrisjsewell — CHANGES_REQUESTED** [review]

File: src/sphinx_book_theme/_transforms.py (hunk 2 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/_transforms.py]: pretty neat. Thanks, have used that now.

**Comment by AakashGfude** [comment]
@choldgraf This looks ready to go 🚀

**Comment by mmcky** [comment]
@AakashGfude this is looking great. Thanks for working on this. 

Will this get "inherited" by any downstream themes such as `quantecon-book-theme`? 

> Apart from mentioning pandoc-sidenote. Thought you would be a better person to put it in the docs.

Do you mind giving this a shot to complete this PR. I can pitch in with some suggestions if needed -- feel free to slack/zoom.

**Comment by AakashGfude** [comment]
> Will this get "inherited" by any downstream themes such as `quantecon-book-theme`?
 yes, it will. 

> Do you mind giving this a shot to complete this PR. I can pitch in with some suggestions if needed -- feel free to slack/zoom.

I have done a futile attempt here https://github.com/executablebooks/sphinx-book-theme/pull/546/commits/8cc71d9e1be5e5c6d1ba0cef2f82a9dd84c450e0

File: docs/content-blocks.md [hunk 3]
```diff
@@ -85,32 +85,87 @@ but I'll stop here.
 ````
 `````
 
-## Sidebars
+(margin:sidenote)=
+## Sidenotes and marginnotes
+
+This theme has support for [Tufte-stype margin / side notes](https://edwardtufte.github.io/tufte-css/). The UX of which has been inspired by [pandoc-sidenote](https://github.com/jez/pandoc-sidenote)
```
Comment (choldgraf) [inline docs/content-blocks.md:91]:
```suggestion
This theme has support for [Tufte-stype margin / side notes](https://edwardtufte.github.io/tufte-css/), with a UX similar to [pandoc-sidenote](https://github.com/jez/pandoc-sidenote).
```

**Comment by AakashGfude** [comment]
@choldgraf should we merge this?

File: docs/content-blocks.md [hunk 4]
```diff
@@ -85,32 +85,87 @@ but I'll stop here.
 ````
 `````
 
-## Sidebars
+(margin:sidenote)=
+## Sidenotes and marginnotes
+
+This theme has support for [Tufte-stype margin / side notes](https://edwardtufte.github.io/tufte-css/), with a UX similar to [pandoc-sidenote](https://github.com/jez/pandoc-sidenote).
```
Comment (choldgraf) [inline docs/content-blocks.md:91]:
```suggestion
This theme has support for [Tufte-style margin / side notes](https://edwardtufte.github.io/tufte-css/), with a UX similar to [pandoc-sidenote](https://github.com/jez/pandoc-sidenote).
```

File: docs/content-blocks.md [hunk 5]
```diff
@@ -85,32 +85,87 @@ but I'll stop here.
...
+
+    On narrow screens, this text won't show up unless you click the superscript number!
+
+Marginnotes are not numbered, but behave the same way as sidenotes.
+On mobile devices a symbol will be displayed that will display the marginnote when clicked[^exmn].
+For example, there's a marginnote in the previous sentence, and you should see a symbol show to display it on mobile screens.
+
+[^exmn]: {-} This is a margin note. Notice there isn’t a number preceding the note.
+
+:::{seealso}
```
Comment (choldgraf) [inline docs/content-blocks.md:108]: I noticed that sidenotes and marginnotes don't work if you put them inside of admonitions. They get placed _inside_ the admonition rather than in the margin. Is this a relatively easy fix? If not then we could add a note that says this isn't possible and try fixing it in a follow-up PR.

File: docs/reference/special-theme-elements.md [hunk 6]
```diff
@@ -271,6 +270,17 @@ alt: My figure text
...
+
+[^sn1]: Test sidenote 1.
+[^sn2]: Test sidenote 2.
+
+Here's a sentence[^mn1] with multiple marginnotes[^mn2].
+
+[^mn1]: {-} Test marginnote 1.
+[^mn2]: {-} Test marginnote 2.
 
 ## Nested admonitions
```
Comment (choldgraf) [inline docs/reference/special-theme-elements.md:285]:
```suggestion
[^mn1]: {-} Test marginnote 1.
[^mn2]: {-} Test marginnote 2.


Sidenotes inside of admonitions should behave the same:

:::{note}
An admonition with a sidenote defined in the admonition[^snam1] and another defined outside of the admonition [^snam2].

[^snam1]: Sidenote defined in the admonition.

:::

[^snam2]: Sidenote defined outside the admonition.


## Nested admonitions
```

Comment (choldgraf) [reply docs/reference/special-theme-elements.md:285]: Here's the text I tried to test the buggy behavior when you define sidenotes in an admonition

File: src/sphinx_book_theme/assets/styles/content/_margin.scss [hunk 7]
```diff
@@ -38,6 +38,48 @@ $content-fullwidth-width: percentage(100% / $content-max-width);
   }
 }
 
+label.margin-toggle {
```
Comment (choldgraf) [inline src/sphinx_book_theme/assets/styles/content/_margin.scss:41]: Can we add a `user-select: none` somewhere for the superscript number?

https://css-tricks.com/almanac/properties/u/user-select/

File: src/sphinx_book_theme/_transforms.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+
+class HandleFootnoteTransform(SphinxPostTransform):
+
+    default_priority = 1
+    formats = ("html",)
+
+    def run(self, **kwargs: Any) -> None:
+        theme_options = self.env.config.html_theme_options
+        if theme_options.get("use_sidenotes", False) is True:
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:14]: To remove an extra indendation, could this instead be:

```python
if theme_options.get("use_sidenotes", False) is False:
    return None
...and then continue the logic w/o indentation here
```

File: src/sphinx_book_theme/_transforms.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+                        == node.attributes["ids"][0]
+                    ):
+                        parent = ftnode.parent
+                        text = ftnode.children[1].astext()
+
+                        sidenote = SideNoteNode()
+                        para = docutil_nodes.inline()
+                        label = ftnode.children[0].astext()
+                        if text.startswith("{-}"):
+                            # marginnotes
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:31]:
```suggestion
                            # marginnotes will have content starting with {-}
                            # remove the number so it doesn't show
```

File: src/sphinx_book_theme/_transforms.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+                            para.attributes["classes"].append("marginnote")
+                            para.append(docutil_nodes.Text(text.replace("{-}", "")))
+
+                            sidenote.attributes["names"].append(
+                                f"marginnote-role-{label}"
+                            )
+                        else:
+                            # sidenotes
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:39]:
```suggestion
                            # sidenotes are the default behavior if no {-}
                            # in this case we keep the number
```

File: src/sphinx_book_theme/_transforms.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+                            superscript = docutil_nodes.superscript("", label)
+                            para.attributes["classes"].append("sidenote")
+                            para.extend([superscript, docutil_nodes.Text(text)])
+
+                            sidenote.attributes["names"].append(
+                                f"sidenote-role-{label}"
+                            )
+                            sidenote.append(superscript)
+                        node.replace_self([sidenote, para])
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:48]: I guess somewhere around here we'd need to look for a parent block and place the `para` in the parent instead of right next to the sidenote? Feels like this might be too much complexity to worry about this edge-case...

**Review by choldgraf — APPROVED** [review]
A few more quick comments, but I think we can try to resolve and merge quickly. If anything seems like it'll take a while, let's just merge it and fix in a later PR. I'm +1 to merge after comments addressed

**Comment by choldgraf** [comment]
Friendly ping here @AakashGfude - let me know if you don't have time for another cycle here, or if you'd like to give this another quick push so we can merge

**Comment by AakashGfude** [comment]
> Friendly ping here @AakashGfude - let me know if you don't have time for another cycle here, or if you'd like to give this another quick push so we can merge

Hi, @choldgraf will review this tomorrow, was traveling for a few days.

**Comment by choldgraf** [comment]
@AakashGfude no worries - I just wanted to see if you still had the bandwidth for another cycle or if you wanted to spot-check these things in future PRs. I think anything that turns out to be complex doesn't need to be a blocker on this PR, but figured I'd give a final round of feedback in case it helps :-)

**Comment by AakashGfude** [comment]
> @AakashGfude no worries - I just wanted to see if you still had the bandwidth for another cycle or if you wanted to spot-check these things in future PRs. I think anything that turns out to be complex doesn't need to be a blocker on this PR, but figured I'd give a final round of feedback in case it helps :-)

Yes for sure, thanks for taking out the time to suggest further helpful improvements. Really appreciate it :)

**Comment by choldgraf** [comment]
I think for the "nested footnotes" bug, one hacky (but maybe reasonable?) way to solve this would be to do something like:

- Move the footnote content next to the footnote reference, like we do now
- Check whether the *parent* of this location is the document. If not, then the footnote content to *just before* the parent node.
- Repeat this process until the footnote content's parent is now the document itself

File: src/sphinx_book_theme/assets/styles/content/_margin.scss (hunk 7 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/assets/styles/content/_margin.scss:40]: Added here https://github.com/executablebooks/sphinx-book-theme/pull/546/commits/75d40af99d42133cd7d5c9fd70a14518bf2eea20

File: src/sphinx_book_theme/_transforms.py (hunk 8 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/_transforms.py]: sure thing. Done in https://github.com/executablebooks/sphinx-book-theme/pull/546/commits/bb485b6ccbce37083a0df073f1a174e380b3a756

**Comment by AakashGfude** [comment]
Thanks for the commits and the suggestions @choldgraf.
And also for putting forward your suggestion for the algo,i.e., nested sidenotes/margin notes. Have implemented a basic one in https://github.com/executablebooks/sphinx-book-theme/pull/546/commits/ac9dcabf10015fb7feea6b159d9aaf8879efa81e. Let me know if you can think of edge cases. I will write a test case for it tomorrow morning, and check for any improvements.

**Comment by choldgraf** [comment]
It looks good! One UX bug I noticed is that on narrow screens for a side/marginnote inside an admonition, clicking it doesn't seem to have an effect

**Comment by AakashGfude** [comment]
> It looks good! One UX bug I noticed is that on narrow screens for a side/marginnote inside an admonition, clicking it doesn't seem to have an effect.

Yeah, this is a tricky one, as we have placed the text outside the admonition. Will have to think of something.

**Comment by AakashGfude** [comment]
> It looks good! One UX bug I noticed is that on narrow screens for a side/marginnote inside an admonition, clicking it doesn't seem to have an effect.

@choldgraf has resolved it, by keeping two copies of the `side/marginnote` text. One outside admonition as discussed. And one in the original place which remains hidden when on a desktop. When using mobile, the hidden one is visible. 


Screen Shot 2022-05-13 at 11 33 53 am

What do you think?

File: src/sphinx_book_theme/assets/styles/base/_base.scss [hunk 12]
```diff
@@ -10,6 +10,10 @@
   left: 0;
 }
 
+.d-n {
```
Comment (choldgraf) [inline src/sphinx_book_theme/assets/styles/base/_base.scss:13]: Can you just use `d-none` for now since we are inheriting bootstrap's classes anyway?

File: src/sphinx_book_theme/_transforms.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+from docutils import nodes as docutil_nodes
+from sphinx import addnodes as sphinx_nodes
+from .nodes import SideNoteNode
+import copy
+
+
+class HandleFootnoteTransform(SphinxPostTransform):
+
+    default_priority = 1
+    formats = ("html",)
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:12]:
```suggestion
class HandleFootnoteTransform(SphinxPostTransform):    
    """Transform footnotes into side/marginnotes."""    
    default_priority = 1
    formats = ("html",)
```

File: src/sphinx_book_theme/_transforms.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+    def run(self, **kwargs: Any) -> None:
+        theme_options = self.env.config.html_theme_options
+        if theme_options.get("use_sidenotes", False) is False:
+            return None
+        for node in self.document.traverse(docutil_nodes.footnote_reference):
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:18]:
```suggestion
        # Cycle through footnote references, and move their content next to the reference
        # This lets us display the reference in the margin, or just below on narrow screens.
        for node in self.document.traverse(docutil_nodes.footnote_reference):
```

File: src/sphinx_book_theme/_transforms.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+            parent = None
+            for ftnode in self.document.traverse(docutil_nodes.footnote):
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:20]:
```suggestion
            # Each footnote reference should have a single node it points to via `ids`
            for ftnode in self.document.traverse(docutil_nodes.footnote):
```

File: src/sphinx_book_theme/_transforms.py (hunk 14 above)
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:18]: Can you use a more descriptive word than `node`? How about `ref_node` and below `foot_node`.

File: src/sphinx_book_theme/_transforms.py [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+                # matching the footnote reference with footnote
+                if (
+                    len(ftnode.attributes["backrefs"])
+                    and ftnode.attributes["backrefs"][0] == node.attributes["ids"][0]
+                ):
+                    parent = ftnode.parent
+                    text = ftnode.children[1].astext()
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:27]: Explain why we do `children[1]`. What is the first element of `children`? What is the element we want to grab? This can be helpful in case docutils changes this ordering in the future

File: src/sphinx_book_theme/_transforms.py [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+
+                    sidenote = SideNoteNode()
+                    para = docutil_nodes.inline()
+                    label = ftnode.children[0].astext()
+
+                    if text.startswith("{-}"):
+                        # marginnotes will have content starting with {-}
+                        # remove the number so it doesn't show
+                        para.attributes["classes"].append("marginnote")
+                        para.append(docutil_nodes.Text(text.replace("{-}", "")))
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:37]: Should we do a `.strip()` here as well since there will be a space after the `{-}`?

File: src/sphinx_book_theme/_transforms.py [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+                        # sidenotes are the default behavior if no {-}
+                        # in this case we keep the number
+                        superscript = docutil_nodes.superscript("", label)
+                        para.attributes["classes"].append("sidenote")
+                        para.extend([superscript, docutil_nodes.Text(text)])
+
+                        sidenote.attributes["names"].append(f"sidenote-role-{label}")
+                        sidenote.append(superscript)
+
+                    # for nested footnotes/marginnotes
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:50]:
```suggestion
                    # If the reference is nested (e.g. in an admonition), duplicate the content node
                    # And place it just before the parent container, so it works w/ margin
                    # Only show one or another depending on screen width.
```

File: src/sphinx_book_theme/_transforms.py [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+                        if not isinstance(
+                            node_parent,
+                            (docutil_nodes.paragraph, docutil_nodes.footnote),
+                        ):
+                            node_parent.replace_self([para, node_parent])
+                            para_dup.attributes["classes"].append("d-n")
+                            break
+                        node_parent = node_parent.parent
+
+                    node.replace_self([sidenote, para_dup])
```
Comment (choldgraf) [inline src/sphinx_book_theme/_transforms.py:67]: Rather than `para_dup` can we be more specific? Like `para_nested` and `para_margin`?

File: src/sphinx_book_theme/assets/styles/content/_margin.scss [hunk 20]
```diff
@@ -38,6 +37,54 @@ $content-fullwidth-width: percentage(100% / $content-max-width);
...
```
Comment (choldgraf) [inline src/sphinx_book_theme/assets/styles/content/_margin.scss:40]:
```suggestion
// Sidenotes and marginnotes
label.margin-toggle {
```

File: src/sphinx_book_theme/nodes.py [hunk 21]
```diff
@@ -0,0 +1,36 @@
+from docutils import nodes
+from sphinx.application import Sphinx
+from typing import Any, cast
+
+
+class SideNoteNode(nodes.Element):
+    """A node that will not be rendered."""
```
Comment (choldgraf) [inline src/sphinx_book_theme/nodes.py:7]: This doesn't sound like the correct docstring

**Review by choldgraf — COMMENTED** [review]
Thanks for these updates @AakashGfude - I've added a few suggestions and comments in there, mostly around trying to make it explicit what the code is doing so that it is more maintainable. Let me know what you think!

File: src/sphinx_book_theme/assets/styles/base/_base.scss (hunk 12 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/assets/styles/base/_base.scss:13]: sure thing

File: src/sphinx_book_theme/_transforms.py (hunk 17 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/_transforms.py:41]: We can I guess. But without `strip()` it seems to be stripping spaces as well, no matter how much space we give.

File: src/sphinx_book_theme/_transforms.py (hunk 14 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/_transforms.py:21]: sounds good.

File: src/sphinx_book_theme/nodes.py (hunk 21 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/nodes.py:7]: Oops.

File: src/sphinx_book_theme/_transforms.py (hunk 19 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/_transforms.py:73]: @choldgraf The variable itself is just a duplicate of para. So, I thought the name should just signify that without adding any extra connotations. Else I can use `para_nested` probably?

File: src/sphinx_book_theme/nodes.py (hunk 21 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/nodes.py]: done in [a87245f](https://github.com/executablebooks/sphinx-book-theme/pull/546/commits/a87245f0706bbf8c070a78a6329783289cc7f1e1)

File: src/sphinx_book_theme/_transforms.py (hunk 16 above)
Comment (AakashGfude) [reply src/sphinx_book_theme/_transforms.py]: done in [f5efddf](https://github.com/executablebooks/sphinx-book-theme/pull/546/commits/f5efddf4772fa3d2f39b3755ecabe5d228b90e54)

File: src/sphinx_book_theme/_transforms.py (hunk 17 above)
Comment (choldgraf) [reply src/sphinx_book_theme/_transforms.py:45]: fair enough - maybe this is a sphinx parsing thing under the hood

File: src/sphinx_book_theme/assets/styles/base/_base.scss [hunk 22]
```diff
@@ -10,6 +10,10 @@
   left: 0;
 }
 
+.d-none {
+  display: none;
+}
+
```
Comment (choldgraf) [inline src/sphinx_book_theme/assets/styles/base/_base.scss:16]:
```suggestion
```

My point was that I don't think we need this at all, because bootstrap should already give it to us

Comment (AakashGfude) [reply src/sphinx_book_theme/assets/styles/base/_base.scss]: aaah yeah. ofcourse.

**Comment by choldgraf** [comment]
OK I think I fixed up the docs conflict, wanna take a look and let me know if it reads OK to you?

Other than that, anything else you'd like to tackle here?

**Comment by AakashGfude** [comment]
@choldgraf the doc looks good to me. Thank you! 
I could not use the `d-none` class instead of `d-n` like you mentioned here https://github.com/executablebooks/sphinx-book-theme/pull/546#discussion_r872050295. As `d-none` class in pydata-sphinx-theme has `display:none !important` in it, which is preventing it from being overridden. 
 
Else I think rest is good.

File: src/sphinx_book_theme/assets/styles/base/_base.scss [hunk 23]
```diff
@@ -23,6 +23,10 @@
   width: 1px !important;
 }
 
+.d-n {
```
Comment (choldgraf) [inline src/sphinx_book_theme/assets/styles/base/_base.scss:26]:
```suggestion
// We need to define our own display-none class since bootstrap uses !important and we want to be able to over-ride
.d-n {
```

**Review by choldgraf — APPROVED** [review]
OK this looks good to me - let's merge it in and see how folks like using it, and we can iterate from there!

**Comment by choldgraf** [comment]
Thanks very much @AakashGfude for this addition!

**Comment by AakashGfude** [comment]
Thank you, @choldgraf.
